e2e/glustershd_test.go

```diff
@@ -64,11 +57,126 @@ func TestSelfHeal(t *testing.T) {
+	optionReq.Advanced = true
+
+	r.Nil(client.VolumeSet(vol1.Name, optionReq))
+
 	// delete volume
 	r.Nil(client.VolumeDelete(vol1.Name))
 }
+
+func checkForPendingHeals(healInfo *shdapi.BrickHealInfo) error {
+	if *healInfo.EntriesInHealPending != 0 {
+		return errors.New("expecting no pending heals, found pending heals")
+	}
+	return nil
+}
+
+func testGranularEntryHeal(t *testing.T, tc *testCluster) {
+	r := require.New(t)
+
+	var brickPaths []string
+
+	for i := 1; i <= 2; i++ {
+		brickPath := testTempDir(t, "brick")
+		brickPaths = append(brickPaths, brickPath)
+	}
+
+	// create 2x2 dist-rep volume
+	createReq := api.VolCreateReq{
+		Name: volname,
+		Subvols: []api.SubvolReq{
+			{
+				ReplicaCount: 2,
+				Type:         "replicate",
+				Bricks: []api.BrickReq{
+					{PeerID: tc.gds[0].PeerID(), Path: brickPaths[0]},
+					{PeerID: tc.gds[0].PeerID(), Path: brickPaths[1]},
+				},
+			},
+		},
+		Force: true,
+	}
+	_, err := client.VolumeCreate(createReq)
+	r.Nil(err)
+
+	r.Nil(client.VolumeStart(volname, false), "volume start failed")
+
+	healInfo, err := client.SelfHealInfo(volname, "info-summary")
+	r.Nil(err)
+	for node := range healInfo {
+		if healInfo[node].Status == "Connected" {
+			r.Nil(checkForPendingHeals(&healInfo[node]))
+		}
+	}
+
+	var optionReq api.VolOptionReq
+	optionReq.Options = map[string]string{"replicate.granular-entry-heal": "enable"}
+	optionReq.Advanced = true
+	r.Nil(client.VolumeSet(volname, optionReq))
+
+	optionReq.Options = map[string]string{"replicate.self-heal-daemon": "off"}
+	optionReq.Advanced = true
+	r.Nil(client.VolumeSet(volname, optionReq))
+
+	checkFuseAvailable(t)
+
+	mntPath := testTempDir(t, "mnt")
+	defer os.RemoveAll(mntPath)
+
+	host, _, _ := net.SplitHostPort(tc.gds[0].ClientAddress)
+	mntCmd := exec.Command("mount", "-t", "glusterfs", host+":"+volname, mntPath)
+	err = mntCmd.Run()
+	defer exec.Command("umount", "-l", mntPath).Run()
+	r.Nil(err, fmt.Sprintf("mount failed: %s", err))
+
+	getBricksStatus, err := client.BricksStatus(volname)
+	r.Nil(err, fmt.Sprintf("brick status operation failed: %s", err))
+	for brick := range getBricksStatus {
+		if getBricksStatus[brick].Info.PeerID.String() == tc.gds[0].PeerID() {
+			process, err := os.FindProcess(getBricksStatus[brick].Pid)
+			r.Nil(err, fmt.Sprintf("failed to find bricks pid: %s", err))
+			err = process.Signal(syscall.Signal(15))
+			r.Nil(err, fmt.Sprintf("failed to kill bricks: %s", err))
+			break
+		}
+	}
+
+	f, err := os.Create(mntPath + "/file1.txt")
+	r.Nil(err, fmt.Sprintf("file creation failed: %s", err))
+	f.Close()
+
+	healInfo, err = client.SelfHealInfo(volname, "info-summary")
+	r.Nil(err)
+	for node := range healInfo {
+		if healInfo[node].Status == "Connected" {
+			r.NotNil(checkForPendingHeals(&healInfo[node]))
+		}
+	}
+
+	optionReq.Options = map[string]string{"replicate.granular-entry-heal": "disable"}
+	optionReq.Advanced = true
+	r.Nil(client.VolumeSet(volname, optionReq))
```

Comment:
same here, is there a way to check `granular-entry-heal` is disabled?

Reply:
Done